### PR TITLE
[docs][python] Add autosummary

### LINKED ISSUE #1326: [Feature request] Table of contents for python API
**Is your feature request related to a problem? Please describe.**

It's a little difficult to navigate the docs and see what classes exist. At the moment it looks like I need to scroll through the docs.

image

**Describe the solution you'd like**

It would be great if there was a heading level which let me see a listing of classes in a module.

### PR DESCRIPTION
**Pull request by ebezzi**
**Issue and/or context:**
Fixes #1326 

**Changes:**
- Adds `autosummary` to the docs. This creates auto-generated pages for each doc entry (functions, classes etc) and groups them using a hierarchical toctree that shows up in the sidebar

**Notes for Reviewer:**
- Preview here: https://tiledbsoma.readthedocs.io/en/ebezzi-autosummary/

### CODE REVIEW

File: apis/python/src/tiledbsoma/__init__.py [hunk 1]
```diff
@@ -6,7 +6,7 @@
 Python package is an implementation of SOMA using the
 `TileDB Embedded <https://github.com/TileDB-Inc/TileDB>`_ engine.
 
-Provides:
+Provides
```
Comment (ebezzi) [inline apis/python/src/tiledbsoma/__init__.py:9]: Colons should be removed as these end up into toctree entities.

File: doc/source/python-api.rst [hunk 2]
```diff
@@ -18,17 +18,10 @@ Features:
...
-.. automodule:: tiledbsoma.io
-   :members:
-
-The tiledbsoma.logging module
------------------------------
-.. automodule:: tiledbsoma.logging
-   :members:
+
+.. toctree:: 
+   :maxdepth: 2
```
Comment (ebezzi) [inline doc/source/python-api.rst:23]: Can be set to 1 if we don't want this toctree to include the subsections.

File: doc/source/python-tiledbsoma-io.rst [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+    :toctree: _autosummary/
+    :nosignatures:
+
+    tiledbsoma.io.add_X_layer
+    tiledbsoma.io.add_matrix_to_collection
+    tiledbsoma.io.create_from_matrix
+    tiledbsoma.io.from_anndata
+    tiledbsoma.io.from_h5ad
+    tiledbsoma.io.to_anndata
+    tiledbsoma.io.to_h5ad
```
Comment (johnkerl) [inline doc/source/python-tiledbsoma-io.rst:22]: Let's go "top down"

Comment (johnkerl) [inline doc/source/python-tiledbsoma-io.rst:22]:
```suggestion
    tiledbsoma.io.from_h5ad
    tiledbsoma.io.from_anndata
    tiledbsoma.io.add_X_layer
    tiledbsoma.io.add_matrix_to_collection
    tiledbsoma.io.create_from_matrix
    tiledbsoma.io.to_h5ad
    tiledbsoma.io.to_anndata
```

Idea:

`from_h5ad`, `from_anndata`, `to_h5ad`, and `to_anndata` are the "big four"

the rest are helpers for doing things like adding an `obsm` element

File: doc/source/python-tiledbsoma-logging.rst [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+-------
+
+.. autosummary::
+    :toctree: _autosummary/
+    :nosignatures:
+
+    tiledbsoma.logging.debug
+    tiledbsoma.logging.info
+    tiledbsoma.logging.log_io
+    tiledbsoma.logging.warning
```
Comment (johnkerl) [inline doc/source/python-tiledbsoma-logging.rst:19]:
```suggestion
    tiledbsoma.logging.info
    tiledbsoma.logging.debug
    tiledbsoma.logging.warning
    tiledbsoma.logging.log_io
```

the first three are by increasing verbosity (which is not alphabetical)

the fourth one, we should (on this PR or another) mark as non-public b/c it's only meant for use by `tiledbsoma.io`

File: doc/source/python-tiledbsoma.rst [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    tiledbsoma.AxisQuery
+    tiledbsoma.Collection
+    tiledbsoma.DataFrame
+    tiledbsoma.DenseNDArray
+    tiledbsoma.Experiment
+    tiledbsoma.ExperimentAxisQuery
+    tiledbsoma.Measurement   
+    tiledbsoma.ResultOrder
+    tiledbsoma.SOMATileDBContext
+    tiledbsoma.SparseNDArray
```
Comment (johnkerl) [inline doc/source/python-tiledbsoma.rst:25]:
```suggestion
    tiledbsoma.Collection
    tiledbsoma.Experiment
    tiledbsoma.Measurement   

    tiledbsoma.DataFrame
    tiledbsoma.SparseNDArray
    tiledbsoma.DenseNDArray

      tiledbsoma.ResultOrder

    tiledbsoma.AxisColumnNames
    tiledbsoma.AxisQuery
    tiledbsoma.ExperimentAxisQuery
    
    tiledbsoma.SOMATileDBContext
```

the first three are collections

the next three are foundational types

result order is a helper for use by the foundational types' `read`

the next three are for queries

that last one is common to all

File: doc/source/python-tiledbsoma.rst [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+    tiledbsoma.get_implementation
+    tiledbsoma.get_implementation_version
+    tiledbsoma.get_storage_engine
+    tiledbsoma.open
+    tiledbsoma.show_package_versions
+    tiledbsoma.tiledbsoma_stats_disable
+    tiledbsoma.tiledbsoma_stats_dump
+    tiledbsoma.tiledbsoma_stats_enable
+    tiledbsoma.tiledbsoma_stats_reset
```
Comment (johnkerl) [inline doc/source/python-tiledbsoma.rst:52]:
```suggestion
    tiledbsoma.open

    tiledbsoma.show_package_versions

    tiledbsoma.get_implementation
    tiledbsoma.get_implementation_version
    tiledbsoma.get_storage_engine

    tiledbsoma.tiledbsoma_stats_disable
    tiledbsoma.tiledbsoma_stats_dump
    tiledbsoma.tiledbsoma_stats_enable
    tiledbsoma.tiledbsoma_stats_reset
```

**Review by johnkerl — APPROVED** [review]
LGTM, except I think alphabetical ordering isn't ever the intuitive thing to do -- I left concrete suggestions

File: doc/source/python-tiledbsoma-logging.rst (hunk 4 above)
Comment (ebezzi) [reply doc/source/python-tiledbsoma-logging.rst]: I removed it from the doc. Another pro of having explicitly declared methods!

**Pull request merged**